### PR TITLE
Query-sharding: shard `group` aggregator

### PR DESCRIPTION
**Pull request by colega**
#### What this PR does

Some customers want to count their series, and with adaptive metrics on Grafana Cloud we suggest them to use the `group by ()` operator. However, that operator wasn't previously sharded, so we hit the chunks limit when querying large number of series.

This adds support for query-sharding for the group by operator.

#### Which issue(s) this PR fixes or relates to

Fixes #<issue number>

#### Checklist

- [x] Tests updated
- [ ] Documentation added
- [x] `CHANGELOG.md` updated - the order of entries should be `[CHANGE]`, `[FEATURE]`, `[ENHANCEMENT]`, `[BUGFIX]`

### CODE REVIEW

File: pkg/frontend/querymiddleware/querysharding_test.go [hunk 1]
```diff
@@ -182,6 +182,18 @@ func TestQuerySharding_Correctness(t *testing.T) {
...
+			expectedShardedQueries: 1,
+		},
+		`group by (group1) (group by (group_1, group_2) (metric_counter))`: {
+			query:                  `group by (group1) (group by (group_1, group_2) (metric_counter))`,
+			expectedShardedQueries: 1,
+		},
+		`count by (group1) (group by (group_1, group_2) (metric_counter))`: {
+			query:                  `count by (group1) (group by (group_1, group_2) (metric_counter))`,
+			expectedShardedQueries: 1,
+		},
```
Comment (dimitarvdimitrov) [inline pkg/frontend/querymiddleware/querysharding_test.go:196]: group1 -> group_1?

Comment (colega) [reply pkg/frontend/querymiddleware/querysharding_test.go]: Thanks, good catch. Fixed.

Comment (dimitarvdimitrov) [reply pkg/frontend/querymiddleware/querysharding_test.go]: unfortunately, that's as much as I can review this PR 😢 The rest SGTM, but i don't feel confident stamping an approval

File: pkg/frontend/querymiddleware/astmapper/sharding.go [hunk 2]
```diff
@@ -294,6 +300,44 @@ func (summer *shardSummer) shardAggregate(expr *parser.AggregateExpr) (mapped pa
...
+		  group by(foo) (rate(bar1{__query_shard__="0_of_2",baz="blip"}[1m])) or
+		  group by(foo) (rate(bar1{__query_shard__="1_of_2",baz="blip"}[1m]))
+		)
+
+		parallelizing a non-parameterized group is representable as
+		group (
+		  group (rate(bar1{__query_shard__="0_of_2",baz="blip"}[1m])) or
+		  group (rate(bar1{__query_shard__="1_of_2",baz="blip"}[1m]))
+		)
+	*/
```
[the remark this answers is not in the document]
Comment (colega) [reply pkg/frontend/querymiddleware/astmapper/sharding.go:323]: It's not, it's just trying to explain what the following code does, just like the comment in the function below.

File: CHANGELOG.md [hunk 3]
```diff
@@ -17,6 +17,8 @@
 * [ENHANCEMENT] Distributor: add experimental support for storing metadata when ingesting metrics via OTLP. This makes metrics description and type available when ingesting metrics via OTLP. Enable with `-distributor.enable-otlp-metadata-storage=true`. #5693 #6035
 * [ENHANCEMENT] Ingester: added support for sampling errors, which can be enabled by setting `-ingester.error-sample-rate`. This way each error will be logged once in the configured number of times. All the discarded samples will still be tracked by the `cortex_discarded_samples_total` metric. #5584 #6014
 * [ENHANCEMENT] Ruler: Fetch secrets used to configure TLS on the Alertmanager client from Vault when `-vault.enabled` is true. #5239
+* [ENHANCEMENT] Query-frontend: added query-sharding support for `group by` aggregation queries. #6024
+* [BUGFIX] Ingester: fix spurious `not found` errors on label values API during head compaction. #5957
```
Comment (dimitarvdimitrov) [inline CHANGELOG.md:21]: this looks out of place

**Review by dimitarvdimitrov — APPROVED** [review]
SGTM apart from an extra changelog entry

File: CHANGELOG.md (hunk 3 above)
Comment (colega) [reply CHANGELOG.md]: Fixed, it was pending a merge.

File: pkg/frontend/querymiddleware/astmapper/sharding_test.go [hunk 4]
```diff
@@ -503,6 +503,16 @@ func TestShardSummer(t *testing.T) {
...
+		{
+			in:                     `group by (a, b) (metric)`,
+			out:                    `group by (a, b) (` + concatShards(3, `group by (a, b) (metric{__query_shard__="x_of_y"})`) + `)`,
+			expectedShardedQueries: 3,
+		},
+		{
+			in:                     `count by (a) (group by (a, b) (metric))`,
+			out:                    `count by (a) (group by (a, b) (` + concatShards(3, `group by (a, b) (metric{__query_shard__="x_of_y"})`) + `))`,
+			expectedShardedQueries: 3,
+		},
```
Comment (pracucci) [inline pkg/frontend/querymiddleware/astmapper/sharding_test.go:515]: [nit] Could you also add this test case, please?

```
in:                     `count(group without () ({namespace="foo"}))`,
			out:                    `count(group without() (` + concatShards(3, `group without() ({namespace="foo",__query_shard__="x_of_y"})`) + `))`,
			expectedShardedQueries: 3,
```

**Review by pracucci — APPROVED** [review]
The `group()` of a `group()` still returns the same result (assuming the grouping labels are the same, as implemented here) so from a correctness perspective I don't see any issue.

Privately, you mentioned the query the customer will run will be something like `count(group without () ({...}))`. Given the grouping will be (intentionally) ineffective (doesn't reduce the result set) making such query shardable would just move the load from the querier to the query-frontend (which typically runs with lower resources than querier). For this reason, I'm not sure this change will effectively solve your problem.

File: pkg/frontend/querymiddleware/astmapper/sharding_test.go (hunk 4 above)
Comment (colega) [reply pkg/frontend/querymiddleware/astmapper/sharding_test.go:515]: Added in https://github.com/grafana/mimir/pull/6024/commits/b1f888f392f0b2511b2fca9ca70fceb86a10b36a

**Pull request merged**